Adding driver for AD7193 (Analog Devices 24-bit ADC)

andrasfuchs opened the pull request:
This is the addition of the driver for AD7193 analog-to-digital converter IC with a sample demonstrating both single and continuous conversion modes.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1024) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1024) before we can accept your contribution.<br/><br/>:x: andrasfuchs [sign now](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1024)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/iot?pullRequest=1024) it.</sub>

joperezr commented:
@andrasfuchs thanks a lot for your contribution! Before we go ahead and review this PR, did you check our existing device binding for ADS1115 ADC? seems like a very similar device and instead of adding a new project, I wonder if we could refactor that in some way so that we can have shared common logic with two derived more specific devices for this one and the Ads1115. Do you mind checking the implementation to see if there is any logic in there that you could reuse and could be refactored?

andrasfuchs commented:
@joperezr Yes, I started with the Ads1115 project as an example when I started to work on the AD7193's driver.

I had the impression that all ADCs could have a common base class, because they share many functions, but I'm not experienced enough yet with other ADCs from other manufacturers to make the proper suggestion.

One thing I would definitely miss from the Ads1115 implementation if I had to use it, is the continuous mode support (that I implemented with AD7193), it's usually a much more performant way to measure the values at a high sample rate.

My plan is to implement one more driver for the more advanced [Analog Devices AD7606](https://www.analog.com/media/en/technical-documentation/data-sheets/AD7606B.pdf) in the next few weeks, and I'm sure I will have a better understanding of the commonalities and differences among ADC later on.

Ideally we would have an common interface for them with the basic functions (like single and continuous conversion) to make the life of the end user easier. 

I would also include some metadata for the devices in general, like manufacturer, model name, datasheet URI etc., because that would make a very good base for a catalog later one, where you will have hundreds of devices supported.

pgrawehr commented:
Note that I already prepared an extension for the ADS1115 driver, that (among other stuff) adds continuous support. See PR #915. There's only a small bug in that driver for now, which should be quick to fix (I actually know what needs to be done).

joperezr commented:
> One thing I would definitely miss from the Ads1115 implementation if I had to use it, is the continuous mode support (that I implemented with AD7193), it's usually a much more performant way to measure the values at a high sample rate.

Also one more thing to note (apart from @pgrawehr saying that he already has a PR adding this support) is that it is totally ok to have a base class with common functionality and then having subclasses extending that so you could still add this support only on your device. All that said, if you can't find a good way to have this code refactored and plug it in to our current ADCs, it is fine for now to keep it as is and then refactor it on a future PR.

joperezr commented on src/devices/Ad7193/Ad7193.sln:
```diff
@@ -0,0 +1,79 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio Version 16
```
We Usually don't add sln files to the repo. This might actually change in the future but at least for consistency lets keep this file out from your PR for now.

pgrawehr replied:
Hmmm... I would definitelly keep them for the individual bindings. You'll need them anyway, and it helps keeping things together a lot. Most recent changes kept it IIRC.

joperezr replied:
I don't believe new bindings have added solutions, at least not that I'm aware of, and if there are any is only because we missed them on the Pull Request. You may be right that perhaps adding them is the right thing to do, but I would rather do this as a bulk for all devices in order to always have consistency on the repo rather than needing to special case.

joperezr commented on src/devices/Ad7193/Ad7193.csproj:
```diff
@@ -0,0 +1,37 @@
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <EnableDefaultItems>false</EnableDefaultItems>
+    <Configurations>Windows-Debug;Linux-Debug;Linux-Release;Windows-Release;Debug;Release</Configurations>
```
Project shouldn't require configurations once you have built using build.cmd once, so please go ahead and remove this line.

pgrawehr replied:
@joperezr (for this and the next few comments) That was my change. And it's the only way I found that made the build pass in VS for all platforms. Without this, if you build on Windows, you get an error saying that System.Device.Gpio.dll was not found (it's using a wrong path). 
This is relevant both for running samples on Windows (for bindings where this is possible) as well as to make sure that any unit tests run on all platforms. We have many unit tests that are not executed on windows due to this (although they are technically platform-independent). I'll create separate PR/issues after further investigation, but I've just tried out this concept here as Andras has run into much the same issues as I have.

joperezr replied:
The code for this binding is AnyCPU and is all IL that applies for all platforms, so there is no need to build it for more than one configuration/platform. If you keep it the way we have it for other bindings and make sure you first build using build.cmd/build.sh from the root, then building individual projects from VS should definitely work for all platforms, if that is not the case it is definitely a bug.

pgrawehr replied:
I do think this is a bug/inconsistency in the build environment. I'll come up with a separate ticket on this, including a more extensive explanation of the problem.

pgrawehr replied:
See #1037

joperezr commented on src/devices/Ad7193/Ad7193.csproj:
```diff
@@ -0,0 +1,37 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!--Cases used when running in VS, with the solution config either Windows-Debug or Linux-Debug (or -Release) -->
```
Please also remove all these next lines and just keep the ProjectReference to System.Device.Gpio which is passing in the RuntimeIdentifier as Linux. You can take a look at our other projects to see what do we usually do.

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+    /// </summary>
+    public class Ad7193 : IDisposable
+    {
+        #region Metadata for IDevice
```
We usually don't include the below members for metadata and instead just add them to the Readme file. There is no real need to have them as public API when they are really just info about the device, so they should really instead just be info in a Readme file.

andrasfuchs replied:
Is there any chance that you will include metadata in the classes themselves? I think they are very useful from the user's perspective, just like in the case of any Windows driver.
This would make it possible to compare different devices, ADCs programmatically, or list all supported devices in the framework, programmatically.
I also use KiCad EDA (a PCB designer program), and they store these pieces of information in their files as well, which is very useful.

joperezr replied:
It would be good to have an actual scenario or need for the APIs first before we expose all these methods that we will then have to support in future versions. That said, if you think it is valuable to keep them, I would much rather have just one method called somehting like GetDeviceMetadata() which would in turn return an object that contained all of these properties. That way you don't polute your surface area, and you keep it simple so that it is a class that is easy to  use and interact with without the need to look into samples and such.

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+        /// <summary>
+        /// The list of SPI modes that are valid for this device
+        /// </summary>
+        public const SpiMode ValidSpiModes = SpiMode.Mode3;
```
NIT: these two SPI related properties shouldn't be part of the public API, if you need them to do validation during execution just keep them as private or internal

andrasfuchs replied:
The SPI properties are related to the technical restriction we must apply if we use this particular device, so I think it is very good if we have this information on the class.
On the other hand, I agree that some check should be done relating to these settings in the constructor to check if the SpiDevice in the right mode/at the right frequency.

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+        /// <summary>
+        /// Number of ADCs on the device
+        /// </summary>
+        public const int ADCCount = 1;
```
Same goes for all these, if they are not configurable (properties) which would affect the interaction with the device, lets keep them off from the surface area.

andrasfuchs replied:
Would you be more open to keep metadata in separate class, or do you insist on removing them completely no matter how they are presented?

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+        /// The list of received ADC values
+        /// </summary>
+        // protected ConcurrentQueue<AdcValue> adcValues = new ConcurrentQueue<AdcValue>();
+        protected BlockingCollection<AdcValue> adcValues = new BlockingCollection<AdcValue>();
```
Why protected if there are no subclasses to this type? It is better to have it as private for now, and then if we ever extend this to have subclasses we can change the visibility of the member

andrasfuchs replied:
This is protected, because I read in one of your contribution documents that you prefer to have a base class with the basic functionality of the driver, and let others extend on that class for the more device specific functions by inheriting from the driver. 
In this case it's important for some advanced functions to keep the access to the ADC values, I think, if someone would, let's say, implement the channel-separation/ordering logic for these values, to have separate buffers for every channel while using the multi-channel continuous conversion.

joperezr replied:
I understand that, but seems to be that there are no subclasses today of this base class. My suggestion was that once somebody does indeed add a subclass from this base, then we can change it back to be protected, but there is no need to have it protected today if nobody is using it.

andrasfuchs replied:
Alright, I changed the scope of it to private.

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+        /// <summary>
+        /// The event that is fired every time a new value is received from the ADC
+        /// </summary>
+        public event EventHandler<AdcValueReceivedEventArgs> AdcValueReceived;
```
NIT: should the name of this event be more something like `OnValueChange` or something like that?

andrasfuchs replied:
I don't think so, but I'm not a native speaker, you are probably a better judge of this :)

When the event fires, the old value doesn't change, but a new value got converted and was added to the list of values in the buffer. 
It's more like when in network communications we receive a new packet that we potentially need to process it.

joperezr replied:
I see, my comment wasn't really about using proper english (I'm also not a native speaker 😄), but it was instead more about how EventHandlers traditionally look in C# APIs. They usually have the pattern of "OnSomethingHappened" so that it is clear that this is an eventhandler and that people that want to use it should just append delegates to it. Given your explanation, probably the right name for it should be `OnValueReceived`?

andrasfuchs replied:
Oh, I see, sure! :) I renamed it.

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+        /// <summary>
+        /// The external reference voltage value. The default is 2.5V on REFIN1+ and REFIN1- (on the Digilent Pmod AD5 board)
+        /// </summary>
+        public double VReference { get; set; } = 2.50f;
```
What is the use case of exposing this?

andrasfuchs replied:
The reference voltage can be set up in a way on the PCB, that is comes from an external source, connected to a particular pin on the IC. 
If the user decides to use an external voltage, they need to set this value manually to reflect their design.

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+        {
+            set
+            {
+                _registerCache[(byte)Register.Configuration] &= 0xFF_FFF8;          // keep all bit values except Gain bits
```
NIT: small thing but we usually do something like `.. &= ~0b111;` for cases like this, which makes it more obvious that what we are doing is to flip those last three bits.

andrasfuchs replied:
I didn't know this syntax, thank you!

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+            set
+            {
+                _registerCache[(byte)Register.Configuration] &= 0xFF_FFF8;          // keep all bit values except Gain bits
+                _registerCache[(byte)Register.Configuration] |= (uint)value;
```
It won't matter much but perhaps you want to check that the value is not greater than 0b111 or you could also just AND it with 0b111 as well.

andrasfuchs replied:
Created two new methods for these kind of operations: GetRegisterBits and SetRegisterBits

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+
+        private readonly StringBuilder _sb = new StringBuilder();
+
+        private readonly uint[] _registerCache = { 0x00, 0x080060, 0x000117, 0x000000, 0xa2, 0x00, 0x000000, 0x000000 };
```
What is the point of keeping a register Cache? I'm not saying that is wrong, but usually you really want to go and check it everytime from the device itself instead of using something from the cache.

andrasfuchs replied:
There are some rare cases when the cache is very useful: when we are in continuous mode, the device send the ADC values continuously on the bus. If we don't ask the ADC to include the current status register in the results, we can win a little bandwidth, but we don't know on which channel that particular ADC value was converted on.
If we previously selected only one channel, that's now a problem, we know that all values are from channel x. 
In this case we should not break the flow of the data, just because we want to know the channel number. It's alright, if we get it from cache.
The situation is similar to values like the gain and filter, these values are typically set before the start of the continuous conversion and they remain the same for sure until we stop that conversion.

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+        /// <summary>
+        /// Gain level on the ADC
+        /// </summary>
+        public Gain PGAGain
```
No Getter?

andrasfuchs replied:
It should have one, I agree.

joperezr commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,838 @@
+        /// <summary>
+        /// Human-readable representation of the current status register
+        /// </summary>
+        public string Status
```
Seems like this getter is actually performing a lot of operations so I guess that we should probably move this to be a GetStatus method instead.

andrasfuchs replied:
Status, Mode and Config are all like this, and I was hesitant about their naming as well.
They basically show the register values in a human readable form, they should be named differently.
What do you think? GetStatus or GetStatusText would be more appropriate? 
Or I could even more these into a RegisterToString(Register register) method, and change the Status, Mode and Config properties to uints with their getters and setters.

joperezr commented:
Hey @andrasfuchs I left some initial feedback that should help get your code to a state where it follows the same patterns as the rest of the repo. Take a look at it and let me know if you have any questions.

andrasfuchs commented:
Thank you @joperezr for the feedback, I appreciate it.

andrasfuchs commented:
@joperezr Could you look at this PR again?

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+﻿using System;
```
copyright header missing

andrasfuchs replied:
What is a copyright header?

krwq replied:
See i.e. https://github.com/dotnet/iot/blob/master/src/System.Device.Gpio/System/Device/Pwm/PwmChannel.Linux.cs#L1-L3

We got guidelines from higher up that every `.cs` file should have it

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Ad7193.Metadata;
+using global::Iot.Units;
```
is global:: necessary?

andrasfuchs replied:
Strange, when I removed it in the past it caused a compilation error, but now it looks fine. I removed it.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        /// <summary>
+        /// The list of received ADC values
+        /// </summary>
+        // private ConcurrentQueue<AdcValue> _adcValues = new ConcurrentQueue<AdcValue>();
```
remove commented out code

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+            {
+                return (GetRegisterBits(Register.Configuration, BitMask.ConfigurationU) == 1);
+            }
+
```
remove extra space

andrasfuchs replied:
I can't see extra space here, but I removed the brackets as you asked at other places.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        /// <summary>
+        /// The external reference voltage value. The default is 2.5V on REFIN1+ and REFIN1- (on the Digilent Pmod AD5 board)
+        /// </summary>
+        public double VReference { get; set; } = 2.50f;
```
can we use full name (VoltageReference)?

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        {
+            get
+            {
+                return (GetRegisterBits(Register.Configuration, BitMask.ConfigurationU) == 1);
```
remove redundant brackets around

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+            {
+                return (Gain)(GetRegisterBits(Register.Configuration, BitMask.ConfigurationG));
+            }
+
```
remove extra enter

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        {
+            get
+            {
+                GetRegisterValue(Register.Mode);
```
is the return value ignored intentionally?

andrasfuchs replied:
yes

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        {
+            get
+            {
+                GetRegisterValue(Register.Status);
```
ditto

andrasfuchs replied:
yes

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+            get
+            {
+                GetRegisterValue(Register.Status);
+                return (GetRegisterBits(Register.Mode, BitMask.StatusRDY) != 0b1);
```
remove extra brackets around expression

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+
+            set
+            {
+                if (value)
```
This could be collapsed to:

```csharp
SetRegisterValue(Register.Communications, value ? 0b0101_1100 : 0b0101_1000);
```

andrasfuchs replied:
This is just the question of preference, I think, but true. I changed it.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        {
+            get
+            {
+                return (GetRegisterBits(Register.Mode, BitMask.ModeDAT_STA) == 1);
```
remove extra brackets

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+
+            set
+            {
+                _registerCache[(byte)Register.Mode] = SetRegisterBits(Register.Mode, BitMask.ModeDAT_STA, value ? 1U : 0);
```
can the caching happen directly in SetRegisterBits? how do you know cache is uninitialized vs initialized with the right value?

andrasfuchs replied:
The cache is initialized with the initial values according to the specification of AD7193.
You're right, the caching should happen in the SetRegisterBits().

andrasfuchs replied:
I moved the cache updater logic into the SetRegisterValue() method, which is also called by the SetRegisterBits() method.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+                        ReadADCValue();
+                        samples++;
+                        jitter = (elapsedTicks - (samples * samplePerTicks));
+                        if (JitterCorrection)
```
i have a mixed feelings on adding utilities like that to devices. One one hand they're useful but on the other hand I suspect these might be useful elsewhere as well. Can you please write in the docs of this device more about this so that if I find similar code to be useful in the future for different device we could unify those?

andrasfuchs replied:
Yes, I had mixed feelings as well. It would be great to have a base class for all the ADCs, so the helper methods like these could go there, unifying the way they do the processing.
At this point we don't have a device class hierarchy, so I think it is a good place to have it, for now.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        /// Reads the current ADC result value on the selected channel
+        /// </summary>
+        /// <returns>24-bit raw value of the last ADC result (+ status byte if enabled)</returns>
+        public uint? ReadADCValue()
```
We usually use TryReadValue/TryReadADCValue instead of nullable. Also if there can be multiple types of error and not just "error" you might want to return status or just throw an exception if this is super unlikely

andrasfuchs replied:
I checked and it can be only one general error, if the Status register indicates it.
I need to test if it is common to have a few errors during conversions, I'll do it today.

andrasfuchs replied:
Alright, so in order to get an error, you must misconfigure something, otherwise the conversion goes on without any errors even at the maximum sampling rate.
Because of this, I changed the return value to a non-nullable uint and added a logic to throw an Exception if something goes wrong. It should happen fairly rarely.

andrasfuchs replied:
I did some further testing, and I was wrong: the error bit gets set even if the signal on the ADC channel is out of range. Because of this, I changed the method to TryReadADCValue(out uint value) returning a bool.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        }
+
+        /// <summary>
+        /// Reads a single value on the selected channel
```
can you document how is this different from ReadADCValue

andrasfuchs replied:
Done

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+                voltage = ((double)adcValue / 8388608.0) - 1.0;
+            }
+
+            voltage *= (VReference / pgaGain);
```
remove extra brackets here and everywhere when you spot (i.e. below)

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+            writeBuffer[0] = commandByte;
+
+            byte[] readBuffer = new byte[writeBuffer.Length];
+            lock (_spiTransferLock)
```
usually we make classes not thread safe and the caller can lock them if they want. I haven't seen any thread being spawned in the ctor - might make sense to remove the locks overall since that will match what we do elsewhere

andrasfuchs replied:
I think I had some trouble with threads while reading the value when I used the event handler in the test application, and that's why I decided to use the lock in the driver. 
It would work if I used it in the sample application as well, but then everyone who uses the event-based value acquisition should know that they need to use locks.
This is a safer solution, although I understand it is not critical, since everyone could do the lock manually when they use this driver.
Would you still prefer to remove it from here and let the users do the locking?

krwq replied:
I'd prefer we removed all the threads and helpers which are not essential. This logic should be made on top of ADC. Even though we currently don't have abstraction we could plug that into it's still better to design APIs so that they don't have anything which adds too much complexity

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+
+            _registerCache[registerAddress] = ByteArrayToUInt32(readBuffer);
+
+            // Debug.WriteLine($"Read Register - address: {registerAddress.ToString("X2")}, command: {commandByte.ToString("X2")}, received: {String.Join(' ', readBuffer.Select(x => x.ToString("X2")))}");
```
remove commented out code

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        /// </summary>
+        /// <param name="register">The register you need the text-representation of</param>
+        /// <returns></returns>
+        public string RegisterToString(Register register)
```
most of the devices made Register enum internal, while those are useful when debugging it's more confusing for users who didn't read datasheet

andrasfuchs replied:
The Register enum can't be internal because of the RegisterToString() method that can be used outside of he library. As you said it's useful for debugging, or in cases when the user needs more detailed information about the state of the device.

krwq replied:
I'd prefer we kept the debugging stuff for library implementers not for everyone

joepr25 replied:
I agree with @krwq here, basically we try device bindings to abstract as much as possible everything related to the actual communication between the device as well as the data-sheet contents, because the whole point of them is to have easy methods to use and only exposing what adds functionality of the sensor/device. Having public API like this will just make it harder to use.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        /// </summary>
+        public void Dispose()
+        {
+            if (_spiDevice != null)
```
null check is not needed (since you use `?.` operator below)

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+
+        private uint GetRegisterBits(Register reg, BitMask bitmask)
+        {
+            uint result = _registerCache[(byte)reg] & (uint)bitmask;     // clear all bit(s) except the bit(s) we would like to read
```
comment above

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,855 @@
+        {
+            uint result = _registerCache[(byte)reg] & (uint)bitmask;     // clear all bit(s) except the bit(s) we would like to read
+
+            if (bitmask > 0)
```
I'm having hard time following what this piece of code is doing. Seems to me it will remove the last n bits of the result (with n being number of 0s in the binary representation of the bitmask) but what's not clear is why do we do that. Could you write a comment explaining what's happening here?

andrasfuchs replied:
Many device has registers and the registers are compartmentalized into smaller values.
For example in the case of [AD7193](https://www.analog.com/media/en/technical-documentation/data-sheets/AD7193.pdf), the 8-bit Status register stores variables like the the 1-bit ready, 1-bit error and 4-bit channel index variables.
To extract these values separately I created the GetRegisterBits method, so I can pass a mask indicating the location of the - let's say - ready bit in the register, and the method returns me only 1-bit of information: 1 if the ready bit is set, 0 is it's not set.
I've just made a description for this method, to make this more understandable.

krwq commented:
@andrasfuchs I'm expecting to take a look at this by the end of the week (hopefully tomorrow) - it's already on my TODO list but it's relatively large. Unfortunatelly I'm working on IoT repo only part time 😞

andrasfuchs commented:
@krwq Alright, no worries, I appreciate your help.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// <summary>
+        /// Polarity select bit: true is unipolar, false is bipolar operation.
+        /// </summary>
+        public bool Unipolar
```
it would be preferred to create an enum `Polarity` with both values although bool is alright too

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// <summary>
+        /// The value of the current Status register
+        /// </summary>
+        public uint Status
```
This should be an enum or API should be internal if it's used anywhere. uint requires users to look in the spec for status codes

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// <summary>
+        /// The value of the current Mode register
+        /// </summary>
+        public uint Mode
```
enum please

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// <summary>
+        /// The value of the current Configuration register
+        /// </summary>
+        public uint Config
```
enum or internal

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// </summary>
+        public bool IsPowerDown
+        {
+            get
```
no setter?

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+            }
+        }
+
+        private bool _continuousRead = false;
```
all fields should be in the beginning of the class

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// <summary>
+        /// Changes the currently selected channel
+        /// </summary>
+        public Channel ActiveChannels
```
Should this be called `SelectedChannel` instead?

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// <summary>
+        /// Set the filter output data rate select bits. The 10 bits of data programmed into these bits determine the filter cutoff frequency, the position of the first notch of the filter, and the output data rate for the part.
+        /// </summary>
+        public ushort Filter
```
Please rename it to Low/HighPassFilterHz and make units clear. If unit is not Hz or multiple it would be better if you did translation here

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        }
+
+        /// <summary>
+        /// The value of the offset calibration register
```
Is there a section in Readme explaining how to do it and what this actually means? Please make the units clear and ideally rename property so it has `Calibration` somewhere in there

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// <summary>
+        /// The value of the full-scale calibration register
+        /// </summary>
+        public uint FullScale
```
same comment as for Offset

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// <returns></returns>
+        public static IDeviceMetadata GetDeviceMetadata()
+        {
+            return Ad7193.metadata;
```
`metadata` should be Capitalized if it's public/protected or `_underscored` if it's private.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        {
+            for (int i = 0; i < 6; i++)
+            {
+                _spiDevice.Write(new byte[] { 0xFF });
```
2 requests:
- can this be done by writing 6 bytes in 1 write instead
- please change to use Span. I.e.:

```csharp
public void Reset()
{
    Span<byte> bytes = stackalloc byte[6];
    bytes.Fill(0xFF); // or just use it as an array, i.e. bytes[0] = 0xFF;
    // ...
}
```

this way you don't allocate 6 arrays on this call

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// <summary>
+        /// Initiates internal calibration, including zero-scale and full-scale calibrations
+        /// </summary>
+        public void Calibrate()
```
would it make sense to make calibration take offset and full scale instead of having two properties? This way it's clearer what those are for. Most of the time you set both fields anyway when you need to calibrate

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+            while (!IsReady)
+            {
+                ReadRegisterValue(Register.Status);
+                Thread.Sleep(5);
```
where is the sleep value coming from? Is the number documented in the spec? If so please mention the page

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        public void StartSingleConversion()
+        {
+            // set the single conversion mode bits
+            SetRegisterBits(Register.Mode, BitMask.ModeMD, 0b001);
```
what if you're in the middle of conversion already and call this?

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+                long elapsedTicks = 0;
+                long jitter = 0;
+                long maxJitterCorrectionPerSample = Math.Max(Stopwatch.Frequency / 100000, 1);
+                while (_stopWatch.IsRunning)
```
How does user get value here? How does it know it got updated? How does user stop this loop? Should you create an event which you call in here? I'd prefer we didn't have any threads in the binding - it's usually better to add helper methods after the binding is complete

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+            {
+                voltage = (double)adcValue / 16777216.0;
+            }
+            else if (!Unipolar)
```
```suggestion
            else
```

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        }
+
+        /// <summary>
+        /// Converts the raw ADC result to Volts
```
Please mention the range depending on Uni/BiPolar setting

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+        /// </summary>
+        /// <param name="adcValue">The raw ADC result</param>
+        /// <returns></returns>
+        protected Temperature ADCValueToTemperature(ulong adcValue)
```
if this is for analog thermometers then please remove the helper. We should decide on the helpers once we got abstraction in place

andrasfuchs replied:
The thermometer is part of the AD7193 IC.

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+
+        private static int GetCommAddress(int x)
+        {
+            return ((x) & 0x07) << 3;
```
brackets around x are not necessary

krwq commented on src/devices/Ad7193/Ad7193.cs:
```diff
@@ -0,0 +1,842 @@
+            return result;
+        }
+
+        private uint ByteArrayToUInt32(byte[] buffer)
```
please remove those helpers and use BinaryPrimitives instead which have utilities for all these

krwq commented on src/devices/Ad7193/AdcValue.cs:
```diff
@@ -0,0 +1,37 @@
+﻿namespace Iot.Device.Ad7193
```
copyright header

krwq commented on src/devices/Ad7193/AdcValueReceivedEventArgs.cs:
```diff
@@ -0,0 +1,24 @@
+﻿using System;
```
copyright header

krwq commented on src/devices/Ad7193/Channel.cs:
```diff
@@ -0,0 +1,79 @@
+﻿using System;
```
copyright header

krwq commented on src/devices/Ad7193/Channel.cs:
```diff
@@ -0,0 +1,79 @@
+        /// (Pseudo Bit = 0) -> +:AIN1 and -:AIN2
+        /// (Pseudo Bit = 1) -> +:AIN1 and -:AINCOM
+        /// </summary>
+        CH00 = 0b00_0000_0001,
```
Channel0, Channel1 etc.. also for bits I'd be easier to read if you used `1 << 0`, `1 << 1`, `1 << 2`, ...

krwq commented on src/devices/Ad7193/Channel.cs:
```diff
@@ -0,0 +1,79 @@
+        /// (Pseudo Bit = 0) -> +:AIN2 and -:AIN2
+        /// (Pseudo Bit = 1) -> +:AINCOM and -:AINCOM
+        /// </summary>
+        Shrt = 0b10_0000_0000
```
what is short channel?

krwq commented on src/devices/Ad7193/Channel.cs:
```diff
@@ -0,0 +1,79 @@
+        /// <summary>
+        /// Temperature sensor
+        /// </summary>
+        TEMP = 0b01_0000_0000,
```
so it is builtin 😄

krwq commented on src/devices/Ad7193/Gain.cs:
```diff
@@ -0,0 +1,38 @@
+    public enum Gain
+    {
+        /// <summary>
+        /// PGA gain off
```
I'd prefer you labeled that as 1x. `Off` might suggest 0

krwq commented on src/devices/Ad7193/Metadata/Ad7193Metadata.cs:
```diff
@@ -0,0 +1,46 @@
+using System.Device.Spi;
+using System.Text;
+
+namespace Ad7193.Metadata
```
does this need separate namespace?

krwq commented on src/devices/Ad7193/Metadata/Ad7193Metadata.cs:
```diff
@@ -0,0 +1,46 @@
+    public class Ad7193Metadata : IDeviceMetadata, ISpiDeviceMetadata, IAdcDeviceMetadata
+    {
+        /// <inheritdoc/>
+        public string Manufacturer => "Analog Devices";
```
Looking at the content, I'd prefer you put most of those in the readme.

krwq commented on src/devices/Ad7193/Metadata/Ad7193Metadata.cs:
```diff
@@ -0,0 +1,46 @@
+        public string DataSheetURI => "https://www.analog.com/media/en/technical-documentation/data-sheets/AD7193.pdf";
+
+        /// <inheritdoc/>
+        public SpiMode ValidSpiModes => SpiMode.Mode3;
```
You should put all SPI settings directly on the ADC class, i.e.:
```
const int DefaultSpiFrequency =  ...;
const SpiMode DefaultSpiMode = ...;
```

I'm not sure about adding validation for Mode3. We can leave this logic for now but some devices are sometimes embedded in others and some settings occasionally change while device is still the same

krwq commented on src/devices/Ad7193/Metadata/ISpiDeviceMetadata.cs:
```diff
@@ -0,0 +1,23 @@
+using System.Device.Spi;
+using System.Text;
+
+namespace Ad7193.Metadata
```
please remove all interfaces with just one implementation. If you'd like the interface to get added for all ADCs please create separate issue.

krwq commented on src/devices/Ad7193/Register.cs:
```diff
@@ -0,0 +1,53 @@
+﻿namespace Iot.Device.Ad7193
```
copyright header - here and everywhere where it's missing

krwq commented on src/devices/Ad7193/Register.cs:
```diff
@@ -0,0 +1,53 @@
+    /// <summary>
+    /// AD7193 Register Map
+    /// </summary>
+    public enum Register : byte
```
this should be internal

krwq commented on src/devices/Ad7193/category.txt:
```diff
@@ -0,0 +1,3 @@
+adc
+spi
+sigma-delta
```
sigma-delta? please remove spi, spi is currently for things exposing SPI, i.e.: SoftwareSpi or FT4222

andrasfuchs replied:
Sigma-delta is a one of the two common types of ADCs.

krwq commented on src/devices/Ad7193/samples/README.md:
```diff
@@ -0,0 +1,140 @@
+## Result
+You can see that the values are oscillating between 0.0 V and 1.0 V at the rate of 4800 SPS, as expected.
+
+![Results in a PowerShell window](PowerShell_RaspberryPi_Output.png)
```
cannot put comment there:
- please remove the checked in DLL
- please make sure all pictures are less than 100kB (i.e., convert to JPEG, lower quality of JPEG)

andrasfuchs replied:
I reduced the JPG file sizes, but the DLL is crucial for the deployment script to work, so I left it there.

krwq replied:
@andrasfuchs:
- we cannot have dlls in the repo, this is because no one can easily verify fidelity and also adds lots of risk (i.e. what if your machine is infected and the dll is also inadvertently infected - everyone using the script will be infected)
- all scripts should not be in the device samples - you should put them in the documentation with instructions and you can provide a link of where to get the dll not the dll itself, we should not do that per binding and rather globally since those won't differ per binding

andrasfuchs commented:
@krwq Thank you for the feedback!

@krwq, @joperezr Thank you guys for spending time on this PR. At this point I feel like that I can't justify more working hours on this, because it functions well for all my use cases, and I'm already behind on my project, because of the many requested changes on this PR.

If you feel like joining in, and modify the PR according to your suggestions, you are more than welcome, otherwise please leave open the PR for a few more weeks, and if I can spare some time, I will work on it a little more.

krwq commented:
Thanks for letting us know @andrasfuchs. Take your time on that, if anyone needs this binding shortly they can help 😄

krwq commented:
@andrasfuchs there still seem to be some unaddressed comments here

andrasfuchs commented:
> @andrasfuchs there still seem to be some unaddressed comments here

Sorry, I don't think I'll be able to work on this in the near future. 
Feel free to make the suggested modifications or discard the pull request.

joperezr commented:
[Triage] PR seems abandoned and author won't have time to finish it. We will create an issue for this binding and attach the progress made on this PR so if anybody is interested in the future they can build upon the progress that was made here. We will close this PR next week if there is no activity by then.
